lib/plugins/rotate_worker_lib/asg.rb

```diff
@@ -0,0 +1,252 @@
+      end
+    end
+
+    # Re-attach an instance to its ASG.
+    #
+    # @param instance [Aws::AutoScaling::Instance] Instance to re-attach.
+    def reattach_instance(instance)
+      instance.load
+      return unless instance.data.nil? \
+        || %w(Detached Detaching).include?(instance.lifecycle_state)
+
+      until instance.data.nil? || instance.lifecycle_state == 'Detached'
+        sleep 10
+        instance.load
+      end
+      instance.attach
+    end
+
+    # Terminate instances.
+    #
+    # @param instances [Array] (instances for termination)
+    #   List of instances to terminate. Defaults to all instances with outdated
+    #   launch configurations.
+    def terminate_instances(instances = outdated_instances)
+      if instances.any?
+        @step.continue(
+          "Terminating #{instances.size} outdated instances..."
+        )
+      end
+      instances.each do |asg_instance|
+        instance = Aws::EC2::Instance.new(asg_instance.instance_id)
+        begin
+          instance.load
+        rescue Aws::EC2::Errors::InvalidInstanceIDNotFound
+          next
+        end
+
+        next unless %w(stopping stopped).include?(instance.state.name)
+
+        instance.wait_until_stopped
+
+        @step.continue("Terminating #{instance.instance_id}")
+        instance.terminate
+      end
+    end
+
+    def reap_volumes(volumes)
+      volumes.each do |volume_id|
+        begin
+          @step.continue("Deleting volume: #{volume_id}")
+          ec2_client(region: ENV['AWS_REGION'])
+            .delete_volume(volume_id: volume_id)
+        rescue StandardError => e
+          # We're catching all errors here, because failing to reap a volume
+          # is not a critical error, will not cause issues with the release.
+          @step.failure("Failed to delete volume #{volume_id}: #{e.message}")
+        end
+      end
+    end
+
+    # Waits for the ASG to reach the desired capacity.
+    def wait_for_capacity
+      @step.continue(
+        'Replacing outdated instances with new instances for the AutoScaling Group...'
+      )
+      # While we wait for the asg to reach capacity, report instance statuses
+      # to the user.
+      before_wait = proc do
+        instances = []
+        asg.reload.instances.each do |i|
+          instances << " #{i.instance_id} (#{i.lifecycle_state})"
+        end
+
+        @step.continue("Instances: #{instances.join(', ')}")
+      end
+
+      asg.reload.wait_until(before_wait: before_wait, max_attempts: 60,
+                            delay: 30) do |a|
+        instances_up = a.instances.select do |i|
+          i.lifecycle_state == 'InService'
+        end
+        instances_up.length == a.desired_capacity
+      end
+      @step.success('AutoScaling Group up to capacity!')
+    end
+
+    # Shuts down an instance, waiting for the worker to stop processing requests
+    # first. We do this instead of using the API so that
+    # services will be stopped properly.
```

Comment:
Phrases "API" and "worker" are specific to a service, we should remove these.

Reply:
I just realized you use these terminologies in the name of the plugin too, let's remove that as well.

Reply:
Yes makes sense